### PR TITLE
 Implement sendNote for VST2;  Add sendNote stub for Jack

### PR DESCRIPTION
**Pull request by pdesaulniers**
I've tested it, and it seems to work.

### CODE REVIEW

File: distrho/src/DistrhoPluginJack.cpp [hunk 1]
```diff
@@ -512,6 +523,13 @@ class PluginJack
     }
 #endif
 
+#if DISTRHO_PLUGIN_WANT_MIDI_INPUT
```
Comment (falkTX) [inline distrho/src/DistrhoPluginJack.cpp]: sendNoteCallback is only defined if using a plugin with midi input, which means the constructor will fail on a plugin with different configuration :P

File: distrho/src/DistrhoPluginVST.cpp [hunk 2]
```diff
@@ -280,8 +280,28 @@ class UIVst
...
+        midiEvent.midiData[1] = note;
+        midiEvent.midiData[2] = velocity;
+
+        VstEvents event;
+        memset(&event, 0, sizeof(VstEvents));
+
+        event.numEvents = 1;
+        event.events[0] = (VstEvent*)&midiEvent;
+
+        fEffect->dispatcher(fEffect, effProcessEvents, 0, 0, &event, 0.0f);
```
Comment (falkTX) [inline distrho/src/DistrhoPluginVST.cpp]: This is nasty... effProcessEvents is supposed to be called by the plugin and only during processing...
Did you try this with carla? very likely carla will see it being called from the wrong thread.

It needs a different approach I believe.

File: distrho/src/DistrhoPluginJack.cpp (hunk 1 above)
Comment (pdesaulniers) [reply distrho/src/DistrhoPluginJack.cpp]: Oops. That wasn't very good :p

File: distrho/src/DistrhoPluginVST.cpp (hunk 2 above)
Comment (pdesaulniers) [reply distrho/src/DistrhoPluginVST.cpp]: Carla seems happy with it. 

Anyway, what would be a good approach?

Comment (falkTX) [reply distrho/src/DistrhoPluginVST.cpp]: a ring buffer with a read-try-lock on the process/RT side of things.

dpf does not have such thing.
I dont think there will be a lot of events from the UI ever, as RT side can get them pretty quickly.
so perhaps just a static array of 16 events, with a mutex that does tryLock() on vst process would work fine.

**Comment by pdesaulniers** [comment]
Anyway, this is a mess. Let's pretend the pull request never happened :D

**Comment by falkTX** [comment]
You dont have to close it, you can work more on it, and then I squash the entire thing into a single commit. :)

**Pull request closed without merging**